Client/Experiments/Messaging/MessageStore.swift

```diff
@@ -0,0 +1,125 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+/// The Message Store is responible for tracking and updating certain metadata of a Message. This
+/// will primarily help us determine if messages are expired, and perhaps what caused expiry.
+
+protocol MessageStoreProtocol {
+    
+    /// Return associated metadata for preexisting or new messages.
+    func getMessageMetadata(messageId: String) -> MessageMeta
+    
+    /// Track and persist impression counts of the message.
+    func onMessageDisplayed(message: Message)
+    
+    /// Track and persist user interactions with the message.
+    func onMessagePressed(message: Message)
+    
+    /// Do the bookkeeping for message dismissed Counts and expiry.
+    func onMessageDismissed(message: Message)
+    
+}
+
+protocol MessageStoreProvider { }
+
+extension MessageStoreProvider {
+    var messageStore: MessageStore {
+        return MessageStore.shared
+    }
+}
+
+class MessageStore: MessageStoreProtocol {
```

Comment:
```suggestion
class MessagingStore: MessageStoreProtocol {
```
????? I don't know. Maybe? You decide

But definitely prefix with Nimubs